Duplicati/Library/Main/Operation/Backup/BackendUploader.cs

```diff
@@ -179,7 +180,10 @@ public Task Run()
                             }
                             finally
                             {
-                                flush.SetFlushed(lastSize);
+                                if (finishedTask != null && !finishedTask.IsFaulted)
+                                {
+                                    flush.SetFlushed(lastSize);

```

Comment:
I know you're still working on some changes but I wrote up some thoughts.

In my opinion an even easier way of doing this is to change the `FlushBackend` method in the BackupHandler and just remove the `WhenAny` call on the uploader and flush request. await the uploader, that will catch any exceptions and wait for all uploads to be finished, and then await the flush request to get the value.

In some ways I could see both removing the `WhenAny` call and adding this code here. If we encounter an exception uploading a file then we truly haven't flushed the uploads successfully so it makes sense not to set the last volume size. Also in the exception scenario the uploader task is guaranteed to finish first in the `WhenAny` and then we'll just await the uploader anyway, throwing the exception due to the await call.

In the happy path where everything uploads fine the flush task will be guaranteed to finish first in the `WhenAny` and we'll never await the uploader task. It seems safer to me to always await the uploader task to be guaranteed to see any exceptions.

Also, this line here makes me slightly nervous in that if it's never called and someone only awaits the flush request, then they will hang indefinitely. As things are written now I don't think that is possible though. It's kind of a weird situation. There are two tasks for the caller to await, the first one can throw exceptions which they should know about and the second task has the value they want, but the second task can actually be completed before the first task.

Sorry, I can ramble sometimes as I let my thoughts roll out. I hope I was at least partially helpful! :)

Reply:
Thanks @seantempleton.  I like your idea of simplifying the `FlushBackend` method.  I'll look into this some more.

The strange this is, a lot of the failed unit tests are passing for me locally.